### ISSUE TITLE
JAX-based autodiff integration for optimization solvers and parameter identification

### ISSUE THREAD

**Issue by derrring** [derrring/MFG_PDE#129]
## Overview
Add automatic differentiation capabilities to MFGArchon using JAX backend for gradient-based optimization, parameter identification, and sensitivity analysis.

## Context
- **Roadmap**: Phase 3A (Q2 2026 - Optional)
- **Trigger**: User demand OR specific research need
- **Timeline**: 6 weeks | Effort: 25 days
- **Priority**: Low (optional, only if Phase 1 & 2 complete)
- **Current State**: JAX backend exists but underutilized

## Motivation

JAX provides powerful autodiff capabilities that could enable:
1. **Gradient-based MFG solvers**: Solve MFG as optimization problem
2. **Parameter identification**: Inverse problems using gradient descent
3. **Sensitivity analysis**: Compute derivatives of solutions w.r.t. parameters
4. **Neural network training**: Differentiable MFG solvers for learning

### 1. JAX-based Optimization Solvers
```python
import jax
import jax.numpy as jnp

class AutodiffMFGSolver:
    def solve_via_optimization(self, problem: MFGProblem):
        """Solve MFG as optimization problem using autodiff."""
        
        def mfg_objective(u, m):
            """Compute MFG system residual."""
            hjb_residual = compute_hjb_residual(u, m, problem)
            fp_residual = compute_fp_residual(u, m, problem)
            return jnp.sum(hjb_residual**2) + jnp.sum(fp_residual**2)
        
        # Automatic gradient computation
        grad_fn = jax.grad(mfg_objective, argnums=(0, 1))
        
        # Gradient descent or L-BFGS
        (u_star, m_star) = optimize_with_autodiff(
            mfg_objective, grad_fn, initial_guess
        )
        
        return SolverResult(U=u_star, M=m_star, ...)
```

### 2. Parameter Identification
```python
def identify_diffusion_coefficient(observations, initial_guess):
    """Identify σ from observed density evolution."""
    
    def loss(sigma):
        # Solve MFG with current sigma estimate
        problem = MFGProblem(sigma=sigma, ...)
        solution = solve_mfg(problem)
        
        # Compare to observations
        return jnp.sum((solution.M - observations)**2)
    
    # Gradient-based optimization
    grad_fn = jax.grad(loss)
    sigma_optimal = gradient_descent(loss, grad_fn, initial_guess)
    
    return sigma_optimal
```

### 3. Sensitivity Analysis
```python
def compute_sensitivity_to_parameters(problem):
    """Compute ∂u/∂θ and ∂m/∂θ for all parameters θ."""
    
    def solve_parametric(theta):
        problem_theta = update_parameters(problem, theta)
        solution = solve_mfg(problem_theta)
        return solution.U, solution.M
    
    # Automatic Jacobian computation
    jacobian_fn = jax.jacobian(solve_parametric)
    sensitivities = jacobian_fn(problem.parameters)
    
    return sensitivities
```

### Phase 1: Core Infrastructure (2 weeks)
- [ ] Design differentiable MFG residual functions
- [ ] Implement JAX-compatible HJB and FP operators
- [ ] Create gradient checkpointing for memory efficiency
- [ ] Unit tests for gradient correctness

### Phase 2: Optimization Solvers (2 weeks)
- [ ] Implement gradient descent MFG solver
- [ ] Add L-BFGS-B for constrained optimization
- [ ] Benchmark against fixed-point iteration
- [ ] Convergence analysis and validation

### Phase 3: Parameter Identification (1 week)
- [ ] Implement inverse problem framework
- [ ] Add regularization (Tikhonov, TV)
- [ ] Validation with known parameters
- [ ] Examples with noisy observations

### Phase 4: Sensitivity Analysis (1 week)
- [ ] Implement parameter sensitivity computation
- [ ] Add uncertainty quantification tools
- [ ] Visualization of sensitivity fields
- [ ] Documentation and examples

### 1. Inverse Problems
- **Goal**: Identify unknown parameters from observations
- **Example**: Estimate diffusion coefficient from crowd density data
- **Method**: Minimize ||m_predicted - m_observed||² using gradients

### 2. Optimal Control with Complex Constraints
- **Goal**: Solve MFG with non-convex constraints
- **Example**: Network MFG with routing constraints
- **Method**: Use autodiff to handle constraint gradients

### 3. Neural Network Training
- **Goal**: Train neural networks to approximate MFG solutions
- **Example**: Physics-informed neural networks (PINNs) for MFG
- **Method**: Backpropagate through differentiable solver

### 4. Uncertainty Quantification
- **Goal**: Quantify sensitivity to parameter uncertainty
- **Example**: How does solution change with ±10% cost variation?
- **Method**: Compute Jacobians w.r.t. all parameters

## Success Metrics
- ✅ JAX-based optimization solvers converge correctly
- ✅ Parameter identification works with <10% error
- ✅ Sensitivity analysis matches finite-difference approximations
- ✅ Examples demonstrating all use cases
- ✅ Documentation and tutorials

## Dependencies
- JAX (already in optional dependencies)
- jaxlib (CPU/GPU support)
- optax (JAX optimization library) - NEW

## Related Issues
- Roadmap: `docs/development/DEVELOPMENT_ROADMAP_2025-Q4.md`
- Issue #123: Rust acceleration (separate feature)

## Triggers for Implementation

**Implement this feature if**:
1. User requests inverse problem capabilities
2. Research project needs parameter identification
3. Teaching demo requires sensitivity analysis
4. PINNs/neural solver research becomes priority

**DO NOT implement if**:
- No specific use case exists
- Fixed-point iteration satisfies all needs
- Finite-difference gradients sufficient

---

**Status**: Open for future consideration (Phase 3 - optional)
**Last Updated**: 2025-10-09
**Decision**: Deferred pending user demand or research justification

**Comment by derrring** [comment]
**Stale issue triage (2026-02-13, #768)**:

Keeping open. This is explicitly scheduled for Phase 3A (Q2 2026, optional) with the precondition "only if Phase 1 & 2 complete." Phase 1 and 2 are still in progress (v0.17.8).

The JAX backend (`backends/jax_backend.py`) exists but is underutilized — this issue remains the correct tracking point for expanding it.

**Next action**: Re-evaluate when Phase 2 completes (estimated v0.18.x).

**Comment by derrring** [comment]
**Update (v0.18.0)**: Package renamed to MFGArchon. Issue body updated. JAX integration remains a valid long-term direction — could connect to weak-GFDM research (automatic differentiation of meshfree operators for sensitivity analysis).
